perun-core/src/main/java/cz/metacentrum/perun/core/impl/UsersManagerImpl.java

```diff
@@ -939,23 +939,42 @@ public boolean userExtSourceExists(PerunSession sess, UserExtSource userExtSourc
 		Utils.notNull(userExtSource.getLogin(), "userExtSource.getLogin");
 		Utils.notNull(userExtSource.getExtSource(), "userExtSource.getExtSource");
 
+		boolean exists = false;
 		try {
+
+			// check by ext identity (login/ext source ID)
 			if (userExtSource.getUserId() >= 0) {
-				return 1 == jdbc.queryForInt("select 1 from user_ext_sources where login_ext=? and ext_sources_id=? and user_id=?",
+				exists = 1 == jdbc.queryForInt("select 1 from user_ext_sources where login_ext=? and ext_sources_id=? and user_id=?",
 						userExtSource.getLogin(), userExtSource.getExtSource().getId(), userExtSource.getUserId());
 			} else {
-				return 1 == jdbc.queryForInt("select 1 from user_ext_sources where login_ext=? and ext_sources_id=?",
+				exists = 1 == jdbc.queryForInt("select 1 from user_ext_sources where login_ext=? and ext_sources_id=?",
 						userExtSource.getLogin(), userExtSource.getExtSource().getId());
 			}
+
+		} catch(EmptyResultDataAccessException ex) {
+			exists = false;
+		} catch(RuntimeException ex) {
+			throw new InternalErrorException(ex);
+		}
+
+		try {
+			// check by UES ID if present in object
+			if (!exists && userExtSource.getId() > 0) {
+				exists = 1 == jdbc.queryForInt("select 1 from user_ext_sources where id=?",
+						userExtSource.getId());
+			}
+
+			return exists;
```

Comment:
I am afraid that this specific implementation will return "true" even if there is userExtSources with wrong ID (id of different user ext source) and correct combination of login and extSourceId. Maybe the best solution will be create two implementation of this method: One checking by id of user ext source, one checking by login and other attributes and then use implementation you need to on the higher level.

Reply:
@stavamichal yes, you are right. But then we probably need to talk about purpose of this method and its current usage in a code. I wanted to avoid creating two similar methods, since then we would need to call both in some cases (and who will remember this while implementing). We usually don't check object on exact match of all params with DB. We usually use only ID if present and this check is used in entry or before new object creation. But we heavily use this method when manipulating users UserExtSources to check, if passed identity is already registered or not. ID is not much relevant (most times not present in passed object).
So creating two methods is a possible solution, but we need to find all usages and properly decide what to do and update javadoc.

Reply:
@licehammer can you please give your opinion on this thing? If you are ok with it, I will approve it at current state.

Reply:
Indeed we are using the check methods to either validate the the ID of the object and also for checking if the object is not in the DB (in cases where the object have to be unique).

Do we have any use-case where you don't want to check other attributes of the object (integrity check) when you have the ID filled?

Anyway, this cannot be merged as it's written right now. The code is very unclear and I'm pretty sure that it can be written in more comprehensive way.